### PR TITLE
UI Tester update - Implement locating a nested UI for a custom style ListEditor

### PR DESCRIPTION
**Pull request by aaronayres35**
Implements the seventh bullet on issue #1149 (Implement locating a nested UI for a custom style ListEditor. Consequently, one should be able to navigate into nested editors inside a nested UI.)

This PR replaces #1172

### CODE REVIEW

File: traitsui/tests/editors/test_list_editor.py
Comment (rahulporuri) [inline traitsui/tests/editors/test_list_editor.py:78]: instead of `"\b\b\b..."` etc, wouldn't it be better to do `"\b"*len("Fields")`?

File: traitsui/tests/editors/test_list_editor.py
Comment (rahulporuri) [inline traitsui/tests/editors/test_list_editor.py:85]: does `\b` need to be a raw string `"r'\b'"`

Comment (aaronayres35) [reply traitsui/tests/editors/test_list_editor.py:85]: This doesn't seem to work either, but with the change from the previous PR #1171 just "\b" will work

File: traitsui/tests/editors/test_list_editor.py
Comment (aaronayres35) [reply traitsui/tests/editors/test_list_editor.py:78]: This does make the reason for the specific number of "\b"s more clear.  However `"\b"*len("Fields") + "David"` no longer feels like a key sequence when you have to like parse the expression mentally.  I don't have a strong preference either way

File: traitsui/testing/tester/qt4/implementation/list_editor.py
Comment (aaronayres35) [inline traitsui/testing/tester/qt4/implementation/list_editor.py:28]: This could probably be pulled out into a separate module eventually like what was done in the draft branch.  I didn't yet as for this pr it felts unnecessary to create a full separate module to house just this function (but I certainly can change that if needed).

File: traitsui/testing/tester/qt4/implementation/list_editor.py
Comment (aaronayres35) [inline traitsui/testing/tester/qt4/implementation/list_editor.py:31]: I have been thinking about how to modify this so that potentially repeated code doesn't need to be repeated.  For example trying to extract some skeleton of this class and keep it in the common_ui_targets module, but I could not decide on a clean way to do so.  The specifics of accessing the nested UI will vary from editor to editor

File: traitsui/tests/editors/test_list_editor.py
Comment (aaronayres35) [inline traitsui/tests/editors/test_list_editor.py:78]: Eventually we will be able to do this with just one KeySequence using '\b' characters, but there is a known issue with that so this works for now.

File: traitsui/testing/tester/qt4/implementation/list_editor.py
Comment (kitchoi) [reply traitsui/testing/tester/qt4/implementation/list_editor.py:28]: I agree this should be pulled out because this logic is actually toolkit agnostic: It deals with the `UI` object. Furthermore the code is repeated - it is a true duplication in this case.
Size of a module is not an issue.

**Comment by aaronayres35** [comment]
This also gives a wx failure on windows:
```
======================================================================
890FAIL: test_locate_element_and_edit (traitsui.tests.editors.test_list_editor.TestCustomListEditor)
891----------------------------------------------------------------------
892Traceback (most recent call last):
893  File "C:\Users\appveyor\.edm\envs\traitsui-test-3.6-wx\lib\site-packages\traitsui\tests\editors\test_list_editor.py", line 84, in test_locate_element_and_edit
894    self.assertEqual(obj.people[7].name, "David")
895AssertionError: 'Fields' != 'David'
896- Fields
897+ David
898
899
```

like issue #1177 
All failures appear to be coming from the same source (I suspect `EmulateKeyPress` in wx helpers.)

File: traitsui/tests/editors/test_list_editor.py
Comment (kitchoi) [inline traitsui/tests/editors/test_list_editor.py:82]: Minor: I'd have wanted to keep these two were it not because of the ongoing issue with KeyClick... Perhaps for the purpose of this PR we can just keep `DisplayedText()` in the next line. That would be enough to verify we have got the right entry.

File: traitsui/tests/editors/test_list_editor.py
Comment (kitchoi) [inline traitsui/tests/editors/test_list_editor.py:72]: Minor: Instead of deepcopy, we can wrap the global list of Person in a function so we always get a new list.
deepcopy does work...but it involves more complexity/logic than necessary.

File: traitsui/tests/editors/test_list_editor.py
Comment (kitchoi) [inline traitsui/tests/editors/test_list_editor.py:86]: Could we have a test to assert the IndexError if we attempt to get an index that is out-of-range?
I believe the error occurs when we call `item.find_by_name`, not when we call `locate(Index(...))`. While it is slightly counterintuitive (one might expect `locate` to raise instead), I think that is fine. The argument for raising later is that the GUI might have changed state between calling `locate` and calling `find_by_name`. So even if `locate` succeeds, that does not guarantee `find_by_name` will succeed.

File: traitsui/testing/tester/qt4/implementation/list_editor.py
Comment (kitchoi) [reply traitsui/testing/tester/qt4/implementation/list_editor.py:19]: Sounds reasonable to me!

**Review by kitchoi — COMMENTED** [review]
LGTM. Some minor suggestions.

File: traitsui/testing/tester/qt4/implementation/list_editor.py
Comment (aaronayres35) [inline traitsui/testing/tester/qt4/implementation/list_editor.py:48]: I feel like this method is unnecessary and just makes code harder to read.  I think we might as well just do `solver=lambda wrapper, location: cls(target=wrapper.target, index=location.index)` in line 65

File: traitsui/tests/editors/test_list_editor.py
Comment (aaronayres35) [reply traitsui/tests/editors/test_list_editor.py:82]: once #1184 is merged, this test should pass.  Will wait to merge this PR until after that has been merged.

Comment (aaronayres35) [reply traitsui/tests/editors/test_list_editor.py:82]: > once #1184 is merged, this test should pass. Will wait to merge this PR until after that has been merged.

It is possible that it would still if the insertion points start incorrectly.  1184 only made the change in default for range editor.  I could see an issue where the backspaces occur before Fields and we get an output of "FieldsDavid".  I am checking this now in a draft PR on a junk branch that I will delete after the CI is finished.  If a fix is needed for the insertion point stuff I will add it to this PR

**Comment by aaronayres35** [comment]
My suspicions in the above discussion were correct.  I added a helper function to set the insertion point the the right not the left end of a textbox on wx.  This feels a bit tangential to this PR ( I can make this change in a separate PR if needed) but without it, this PR will introduce a new test failure.  After #1184 is merged, with this most recent change, there shouldn't be any new errors with this PR as far as I know.

File: traitsui/tests/editors/test_list_editor.py
Comment (kitchoi) [reply traitsui/tests/editors/test_list_editor.py:82]: We don't know how long it'd take for #1184 to be merged. I'd still prefer decoupling this PR from #1184 so we can review and merge it more efficiently. But if you'd like to wait, that's okay for me.

**Review by kitchoi — APPROVED** [review]
LGTM.
One function being overly indented, can be merged after fixing that.

**Pull request merged**
